### DISCUSSION TITLE
Add Barcode scanning for easy item creation using https://www.barcodelookup.com/

### DISCUSSION THREAD

**Discussion by voron69-bit** [sysadminsmedia/homebox#724] [Ideas]
### What is the problem you are trying to solve with this feature?

Vastly improve the speed in which you can add objects to locations. Current workflow is incredibly tedious, and hamstrings the organization process.

### What is the solution you are proposing?

I just downloaded and tested this free app on Android "Barcode Lookup". It is fantastic. It instantly scans barcodes referencing amazon, and so many other databases. 

### Contributions

- [x] I have searched through existing issues and feature requests to see if my idea has already been proposed.
- [ ] If this feature is accepted, I would be willing to help implement and maintain this feature.
- [ ] If this feature is accepted, I'm willing to sponsor the development of this feature.

**Comment by MisterKLI** [comment]
I second that. Using the barcodes on the products would make the whole process of inventory much easier.

**Comment by 101br03k** [comment]
The mentioned website (barcodelookup.com) in the title is not free. A free account good for two weeks and after that only paid plans.

| plan|api calls per month|price per month|
|-----|---|---|
|STARTER |5,000 |$99| 
|ADVANCED| 25,00|$249   |
|PROFESSIONAL|1,000,000 |$499|
|ENTERPRISE|500,000|949 |

Also it seems that all plans are rate limited.

**Comment by voron69-bit** [comment]
There are numerous other sources to use, but barcode scanning is required.

**Comment by tankerkiller125** [comment]
If someone can find a good, free barcode database, or at the bare minimum a barcode database with very generous free limits I'd be open to this.

Honestly what I'd really love to see is some sort of decentralized database of barcodes that people can either use a hosted instance of or run themselves and receive updates from a community of users, but that's extremely wishful thinking on my part.

**Comment by voron69-bit** [comment]
oooooh. that'd be the cats ass.

**Comment by trevermah** [comment]
I've implemented this one and you can get 100 calls a day from this site. 

Use the barcode scanned to look up against the returned json.  It could grab images (for attachments), and all the other details that it responds with.  Multiple responses could have a popup to show the user to select which one they want to add.

https://www.upcitemdb.com/api/explorer#!/lookup/get_trial_lookup

**Comment by voron69-bit** [comment]
YES. Brilliant idea @trevermah ! This would really make this app top tier.

**Comment by voron69-bit** [comment]
> I've implemented this one and you can get 100 calls a day from this site.
> 
> Use the barcode scanned to look up against the returned json. It could grab images (for attachments), and all the other details that it responds with. Multiple responses could have a popup to show the user to select which one they want to add.
> 
> https://www.upcitemdb.com/api/explorer#!/lookup/get_trial_lookup

How is this accessed? There is no barcode scan feature in the new item creation, I just see the button for adding an image still.

**Comment by tankerkiller125** [comment]
@voron69-bit This is not something implemented in Homebox itself (at least not our version). And likely won't exist for some time. Frankly I consider 100 calls a day to be far too low to be useful (especially for someone wanting to inventory their entire home over the course of a weekend). And I can't and won't in good faith suggest that people pay $100/month for access to a reasonable API limit, even if they only paid for one month.

**Comment by voron69-bit** [comment]
Totally agree on that,  but 100 calls per day certainly is better than entering 100 items manually.  I could inventory everything over a month,  and save so much time still.

**Comment by supaeasy** [comment]
How about these suggestions?
https://github.com/sysadminsmedia/homebox/discussions/477#discussion-7830651

**Comment by swc** [comment]
> ... Frankly I consider 100 calls a day to be far too low to be useful (especially for someone wanting to inventory their entire home over the course of a weekend). ...

But... what about those weekends when you decide not to inventory an entire home? 😜 Literally every day after that weekend, 100/day would be plenty for day-to-day use. The perfect should not be the enemy of the good.

For my use case it would be amazing. In addition to maintaining a home inventory, I'm also an ebay seller. Things come in on a daily basis, but never more than 100 in a day. I want to be able to open a box of incoming stock, slap QR codes on everything, and then scan each QR code followed by the item's barcode. It could auto-populate the item name and possibly the image (as well as some other fields like manufacturer). Then I just scan the bin where I'm storing it and I'm done. And I can transform this data into a csv for auto-creation of draft eBay listings. *Chef's kiss*. 

This would also be great for freezer inventory... It would be really easy to scan packaged items before tossing into the deep freeze.

Another way to really stretch the 100/day limit would be maintaining a local cache of barcode lookup data. I tend to stock the same items over and over again and the system shouldn't be calling the API for a barcode that has been scanned before. In my case, this would reduce the API calls by at least 95%.

**Comment by voron69-bit** [comment]
No I TOTALLY agree with you. I'd rather have the 100 calls a day and be able to slowly chip away, then nothing at all.

**Comment by voron69-bit** [comment]
I've added this feature to a program I've been working on, using zxing's open source UPC scanner js https://zxing-js.github.io/library/examples/multi-camera/ and tying it to https://www.upcitemdb.com/. Sure, it only allows 100 api calls a day, but this is a HUGE improvement over manual entry. If you need to scan more than 100 things a day, you can easily wait till the next day to continue. We're all just home users of this, but trying to use it for a business. This definitely needs to be added.

**Comment by joh6nn** [comment]
Looks like there are a number of free databases available:

- https://www.upcindex.com/
- https://www.upc-search.org/
- https://upcdatabase.org/
- https://www.upczilla.com/

Having done exactly no research other than to notice they exist, my guess here is the reason there's more than one is because they all have various gaps and issues. So, probably the best path forward is to implement a generic lookup that lets users pick which backends (paid and/or free) they want to use, query the selected dbs, and then let the user decide which result to use.

**Comment by voron69-bit** [comment]
> Looks like there are a number of free databases available:
> 
> * https://www.upcindex.com/
> * https://www.upc-search.org/
> * https://upcdatabase.org/
> * https://www.upczilla.com/
> 
> Having done exactly no research other than to notice they exist, my guess here is the reason there's more than one is because they all have various gaps and issues. So, probably the best path forward is to implement a generic lookup that lets users pick which backends (paid and/or free) they want to use, query the selected dbs, and then let the user decide which result to use.

Absolutely!

**Comment by tankerkiller125** [comment]
Please note we're converting this issue into a discussion as per our new procedures/policies. You can read more about that in our announcement #712

**Comment by hummelm10** [comment]
The API limits on any of these with APIs wouldn't really be an issue except for the initial inventory set up in which case I can just spread out the unboxing and organizing. I just want to be able to use the UPC data to easily fill in some of the fields as I work through stuff. Once it's set up I'm only adding new items occasionally but never more than 100 in a day.

It also makes the dataset easily more complete for those using this solution to document items for insurance.

**Comment by crumbowl** [comment]
Hey guys, just submitted a PoC for this feature:

https://github.com/sysadminsmedia/homebox/pull/818

It's very dirty but work well and could be nicely extended (I have a lot of ideas)

For the UI workflow, following are a few proposals:

Proposal 1:

1. Go in Scanner, when a barcode is detected, ask the user if he/she wants to create an item.
2. Display a "gallery" of found items on different platforms
3. The user select the product in the list
4. Open the "Create item" dialog, that is now filled

Proposal 2: 
1. Add an entry in the "Create" menu, e.g: "Create from barcode".
2. Open a dialog for scanning the barcode, OR entering manually a barcode
3. Display a "gallery" of found items on different barcode DBs
4. Open the "Create item" dialog, that is now filled

Proposal 3:
1. Open the "Create item" dialog
2. Click on "fetch product info from barcode"
3. Open a dialog for scanning the barcode, OR entering manually a barcode
4. Display a "gallery" of found items on different barcode DBs
5. Click on the item then click ok to auto-fill the "Create item" dialog


Note that in all of those proposal, barcode value, brand and model fields should appear (dynamically?) on the "Create item" dialog.

Cheers!

**Reply by voron69-bit** [reply]
Brilliant! This would be a HUGE upgrade!

**Reply by tankerkiller125** [reply]
I prefer option 1 IF you can determine Asset QR Code vs UPN/UPC codes.

**Reply by tonyaellie** [reply]
I think option 1 is definitely the best option, maybe with a tip somewhere on the createitem page that you can prefill the information from  a barcode using the scanner

**Comment by crumbowl** [comment]
Hello guys!

This is the a little update for those who missed the latest development I've done on this.

Create item modal:

https://github.com/user-attachments/assets/6a8e9ef8-4a63-45a4-b8cb-7d26bdfda947

In the scanner:

![Screenshot From 2025-07-03 00-10-51](https://github.com/user-attachments/assets/586d07fc-be5a-444b-b8b6-ece38ec254ea)

The new dialog for selecting the product (currently 2 online databases are fetched, might be improved by others)

https://github.com/user-attachments/assets/f1f314a4-9afb-47e6-9a24-00aa9d6be942

**Comment by voron69-bit** [comment]
How do we get this added to Homebox?
This looks AMAZING!

**Reply by tankerkiller125** [reply]
Once the PR is reviewed and approved it will be merged. I have quite a bit going on over the next few days though so it might be a bit, and then of course there's the wait for the next release.

**Reply by voron69-bit** [reply]
Freakin awesome though! This is going to a HUGE help!!! You guys are amazing!

**Comment by JingleheimerSE** [comment]
This is amazing to see barcode functionality being added! Cant wait to check out the feature once its merged.

@crumbowl do you know if the barcode will be available through the API as part of your PR?

I've been putting together a prototype hardware barcode scanner that includes a screen and an ESP32 with the idea that I could use the API to do some of what you have now added UI for. I'll likely continue with the project as I think a cheap physical scanner would be ideal for me given I don't always have a phone and navigating through the web UI would take several steps where as being able just scan and tap add or remove quickly would be a requirement to get the family onboard and using the system. 

My current goals for the scanner project are:
- Scanning a barcode displays the item info and quantity if it exists in HomeBox
- Scanning a barcode allows for adding a new item if it doesn't exist.
- Scanning a barcode or QR code and allows quickly adding / removing one or more of that item from HomeBox.
- Support an advanced mode where scanning a barcode would enable specifying a total quantity of the item so that inventory can be synchronized quickly if someone takes a bunch of items from the pantry and doesn't scan them out.
- Keep the total cost of the device below $25 USD so that they can be bought somewhat in mass and left where they are to be used (e.g. one in the store room, one on the freezer door, and one in the pantry)

In order for this device to function though, I'd have to use the API which at this point doesn't expose Custom Fields from what I can tell. I was going to play some games with the ID or the Serial number fields but it would amazing if we could directly access a UPC field via the API.

**Reply by tankerkiller125** [reply]
Custom fields are available through the API (otherwise the frontend wouldn't be able to display them) https://homebox.software/en/api/#/paths/v1-items-id/get is the API endpoint in question, and the "fields" array contains the custom fields if they already exists.

If you need to add custom fields to items that's part of the update endpoint, but I'll need to look up the body for that request because for some reason the API docs didn't generate that.

**Reply by JingleheimerSE** [reply]
Thanks for the confirmation @tankerkiller125!

I was able to see the `fields` data you mentioned after looking more closely at `items` endpoint and which ID it required. I had tried that before but used the Asset ID rather than the GUID ID and got an error. Somehow I stumbled on the `assets` endpoint directly accepting my Asset ID but didn't return all the extra fields.

**Reply by crumbowl** [reply]
@JingleheimerSE Yes, the API takes a Barcode in input, and will provide a BarcodeProduct object, that contains itself a CreateItem object.

You can check files modified by my PR, the API will probably not change until the final merge:

https://github.com/sysadminsmedia/homebox/pull/818/files

This is what is returned by the API I added:

        "repo.BarcodeProduct": {
            "type": "object",
            "properties": {
                "barcode": {
                    "type": "string"
                },
                "imageBase64": {
                    "type": "string"
                },
                "imageURL": {
                    "description": "TODO: add image attachement\nTODO: add asin?",
                    "type": "string"
                },
                "item": {
                    "$ref": "#/definitions/repo.ItemCreate"
                },
                "manufacturer": {
                    "type": "string"
                },
                "modelNumber": {
                    "description": "Identifications",
                    "type": "string"
                },
                "notes": {
                    "description": "Extras",
                    "type": "string"
                },
                "search_engine_name": {
                    "type": "string"
                }
            }
        },

**Comment by supaeasy** [comment]
What happened to the PR #818 about this implementation? I have been monitoring it closely as I am eager to try it out and now it just vanished...

**Reply by tankerkiller125** [reply]
I don't know actually, looking through audit logs everything was fine 6 hours ago, CodeRabbit was commenting and all that. And that's the last trace of it. No audit log for it's deletion or anything of that nature.

Edit: A little further digging, the authors Github account is also "Not Found" I've reached out to the author on Discord to see if they need some assistance of some kind.

**Reply by supaeasy** [reply]
Thank you for investigating into this. To my understanding it was nearly complete and I was waiting eagerly for this feature. Hell, I even bought a Barcode scanner :-(

**Reply by tankerkiller125** [reply]
We're working with the author to see if we can help them get their account back.

Worst case, I'll have them generate old school patch files and we'll get it sorted from there.

**Comment by voron69-bit** [comment]
Yeah I'd like to know as well. This would be a HUGE help, considering the take a photo from the camera when adding a new item is broken again.

**Comment by crumbowl** [comment]
Hey all! Feature is live!

To test the feature, you can use this image (! it's not a stable version! ) ghcr.io/sysadminsmedia/homebox:main

I am currently writing the doc that will land soon in a PR. In order to maximize number of results, you can create an account and generate a token on barcodespider.com, then fill the env variable HBOX_BARCODE_TOKEN_BARCODESPIDER in your dockerfile with the token. Then scan a barcode in the scanner, or go in the create item dialog: 2 new icons are now available!

Not all products are found yet, and there are few enhancements that are on the roadmap. In the meantime, I would be more than happy to get your feedbacks!

Best regards :)

**Reply by supaeasy** [reply]
Works pretty well!

**Reply by voron69-bit** [reply]
100 free lookups is more than enough per day. I'd love to see this also query amazon upcs for products that have an amazon upc tag on. ( Think electronic parts, heat shrink tube, things that are repackaged by amazon and get a amazon upc )

**Reply by crumbowl** [reply]
Do you have any example of how amazon upc tags look like?

**Reply by voron69-bit** [reply]
Here's a perfect example.
<img width="1836" height="1225" alt="example" src="https://github.com/user-attachments/assets/b92181ff-a5c1-4fc1-97c9-a080b003ac0a" />
This UPC is for this product.
https://www.amazon.com/dp/B0CNGM28QW

Amazon does this a ton, especially with objects that don't have...real packaging? Things like resistors, heat shrink tube, motors, 3d printer parts, nuts,bolts, springs, etc.

**Reply by supaeasy** [reply]
I have an improvement suggestion: could it be possible to add a GTIN field to the items and implement a duplicate check? I have a lot of items inside Homebox already and would like to avoid duplicates in my barcode scanning frenzy.  
I have added GTINs (mostly ISBNs for books) as serial numbers already, though I am aware this isn't the optimal field for that.

**Reply by crumbowl** [reply]
Actually I started to work yesterday night on this!

**Reply by JingleheimerSE** [reply]
I've been playing with the new feature a bit and so far things work well as advertised, really great work!

I have a couple of suggestions that would, in my opinion, make UPCs more useful in general:

1. There should be a UPC field stored along with the other information that is visible from the UI and searchable.
    - Having a UPC field would enable searching for items based on a scan or adding a UPC to existing items (maybe in the future there could be a maintenance task to post-populate items with online database info if information is missing).
2. If a UPC is scanned and no item is found online, allow creating an item anyway (with the UPC field suggested above populated)
    - This would streamline creating items from a UPC even if they aren't in the online databases

**Reply by supaeasy** [reply]
> Actually I started to work yesterday night on this!

Please do this. Will be a massive step up!

**Reply by Estofado** [reply]
Hi,

I believe the Amazon barcode displays the “FNSKU” (Fulfillment Network Stock Keeping Unit). This code is an internal Amazon identifier and is specific to each reseller.

Of course, having access to this database would be great. However, I doubt it’s possible to obtain external access to this barcode database.

**Reply by supaeasy** [reply]
> Actually I started to work yesterday night on this!

Hi there did you make any progress on this?

**Comment by tokendad** [comment]
Possible suggestion on the barcode limits that these sites have.   If you are familiar with Sonarr/Radarr..etc.  They use an caching proxy called skyhook.    Its a middle man that pulls the metadata from TVDB, IMDB, and others.  ITs so the program doesn't flag the API limits of these sites. And is also cleans up the data and normalizes it to Sonarr standards.  Skyhook updats its info every 24hrs from the main sites,  but only pulls new or changed data. 


I know its a big change,  but that could be a future enchanment.  A caching proxy server to get around the API limts.

**Reply by hummelm10** [reply]
This would probably violate the ToS of most of the barcode sites that have paid APIs.

**Reply by supaeasy** [reply]
Why though? This would reduce the load

**Reply by voron69-bit** [reply]
Love this idea. We really need to improve this ability.

**Reply by hummelm10** [reply]
> Why though? This would reduce the load

It would be essentially stealing the db from a paid service to avoid having to pay for an account. The other api's you're talking about are already free sources. 

> You agree to not: (a) use any bot, spider, or other automated script, program, software, process, or technology to scrape, data mine, download, copy, extract or otherwise collect content from the Services, except in connection with your use of the API Services; (b) circumvent or reproduce the look and feel of the Services; (c) copy, damage, decompile, disassemble, disrupt, impair, modify or reverse engineer the Services; (d) use the Services to violate any law or regulation in any jurisdiction; or (e) gain, or attempt to gain, unauthorized access to the Services, including through the use of another user’s account or through any method not provided by or approved by Barcode Lookup.

https://www.barcodelookup.com/terms-and-conditions

If anything, the current design might already violate the ToS and users should have to put in their own API key instead of using one embedded in the project.

**Reply by tankerkiller125** [reply]
As far as I'm aware from reviewing the code there are no embedded API tokens, and it's using non-token free limits.

https://github.com/sysadminsmedia/homebox/blob/1088972ff07132f669af2e2e5b6c81ec7be1bcbf/backend/app/api/handlers/v1/v1_ctrl_product_search.go#L128-L130

https://github.com/sysadminsmedia/homebox/blob/1088972ff07132f669af2e2e5b6c81ec7be1bcbf/backend/app/api/handlers/v1/v1_ctrl_product_search.go#L185-L198

The current implementation does not use barcodelookup.com
